Support generic mapping in builtin assertrepr compare

<!--
Thanks for suggesting a feature!

Quick check-list while suggesting features:
-->

#### What's the problem this feature will solve?
<!-- What are you trying to do, that you are unable to achieve with pytest as it currently stands? -->

Having a nicely formatted assert when comparing mappings that are not subclass of dicts.

#### Describe the solution you'd like
<!-- A clear and concise description of what you want to happen. -->

<!-- Provide examples of real-world use cases that this would enable and how it solves the problem described above. -->

I think it's enough to change the check in `_compare_eq_any` and similar that at the moment does `isinstance(x, dict)`  to mapping, since it seems that `_compare_eq_dict` already supports mappings.

#### Alternative Solutions
<!-- Have you tried to workaround the problem using a pytest plugin or other tools? Or a different approach to solving this issue? Please elaborate here. -->

At the moment, I just do `assert dict(d1) == dict(d2)`

#### Additional context
<!-- Add any other context, links, etc. about the feature here. -->

I can't see how this is error prone with mappings while it was already implemented for sequences:

```python
class MySequence:
    def __init__(self, data):
        self.data = data

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return len(self.data)

    def __repr__(self):
        return f"MySequence({self.data})"


def test_foo():
    assert MySequence([1, 2, 3]) == [1, 2, 4]

```

```
tests/foo_test.py::test_foo FAILED                                                                                                                                                                           [100%]

===================================================================================================== FAILURES =====================================================================================================
_____________________________________________________________________________________________________ test_foo _____________________________________________________________________________________________________

    def test_foo():
>       assert MySequence([1, 2, 3]) == [1, 2, 4]
E       assert MySequence([1, 2, 3]) == [1, 2, 4]
E         
E         Full diff:
E         + MySequence([1, 2, 3])
E         - [
E         -     1,
E         -     2,
E         -     4,
E         - ]

tests/foo_test.py:16: AssertionError
============================================================================================= short test summary info ==============================================================================================
FAILED tests/foo_test.py::test_foo - assert MySequence([1, 2, 3]) == [1, 2, 4]
================================================================================================ 1 failed in 0.07s =================================================================================================
```

--- a/AUTHORS
+++ b/AUTHORS
@@ -205,6 +205,7 @@ Hugo van Kemenade
 Hui Wang (coldnight)
 Ian Bicking
 Ian Lesperance
+Ilya Abdolmanafi
 Ilya Konstantinov
 Ionuț Turturică
 Isaac Virshup

--- a/changelog/14461.improvement.rst
+++ b/changelog/14461.improvement.rst
@@ -1,0 +1,1 @@
+Improved assertion failure explanations for equality comparisons between mapping objects that are not ``dict`` instances.

--- a/src/_pytest/assertion/util.py
+++ b/src/_pytest/assertion/util.py
@@ -156,8 +156,8 @@ def istext(x: object) -> TypeGuard[str]:
     return isinstance(x, str)
 
 
-def isdict(x: object) -> TypeGuard[dict[object, object]]:
-    return isinstance(x, dict)
+def ismapping(x: object) -> TypeGuard[Mapping[object, object]]:
+    return isinstance(x, Mapping)
 
 
 def isset(x: object) -> TypeGuard[set[object] | frozenset[object]]:
@@ -323,8 +323,8 @@ def _compare_eq_any(
             explanation = _compare_eq_sequence(left, right, highlighter, verbose)
         elif isset(left) and isset(right):
             explanation = _compare_eq_set(left, right, highlighter, verbose)
-        elif isdict(left) and isdict(right):
-            explanation = _compare_eq_dict(left, right, highlighter, verbose)
+        elif ismapping(left) and ismapping(right):
+            explanation = _compare_eq_mapping(left, right, highlighter, verbose)
 
         if isiterable(left) and isiterable(right):
             expl = _compare_eq_iterable(left, right, highlighter, verbose)
@@ -576,7 +576,7 @@ def _set_one_sided_diff(
     return explanation
 
 
-def _compare_eq_dict(
+def _compare_eq_mapping(
     left: Mapping[object, object],
     right: Mapping[object, object],
     highlighter: _HighlightFunc,

--- a/testing/test_assertion.py
+++ b/testing/test_assertion.py
@@ -899,7 +899,7 @@ class TestAssert_reprcompare:
             def __iter__(self) -> Iterator[str]:
                 return iter(self._values)
 
-            def __len__(self) -> int:
+            def __len__(self) -> int:  # pragma: no cover
                 return len(self._values)
 
             def __repr__(self) -> str:

--- a/testing/test_assertion.py
+++ b/testing/test_assertion.py
@@ -1,6 +1,8 @@
 # mypy: allow-untyped-defs
 from __future__ import annotations
 
+from collections.abc import Iterator
+from collections.abc import Mapping
 from collections.abc import MutableSequence
 import sys
 import textwrap
@@ -884,6 +886,38 @@ class TestAssert_reprcompare:
             "?      ^   ^",
             "+     'c': 2,",
             "  }",
+        ]
+
+    def test_mapping_different_items(self) -> None:
+        class SimpleMapping(Mapping[str, int]):
+            def __init__(self, values: dict[str, int]) -> None:
+                self._values = values
+
+            def __getitem__(self, key: str) -> int:
+                return self._values[key]
+
+            def __iter__(self) -> Iterator[str]:
+                return iter(self._values)
+
+            def __len__(self) -> int:
+                return len(self._values)
+
+            def __repr__(self) -> str:
+                return f"SimpleMapping({self._values!r})"
+
+        lines = callequal(
+            SimpleMapping({"a": 0, "b": 1}),
+            SimpleMapping({"a": 1, "b": 1, "c": 2}),
+        )
+
+        assert lines is not None
+        assert lines[2:] == [
+            "Omitting 1 identical items, use -vv to show",
+            "Differing items:",
+            "{'a': 0} != {'a': 1}",
+            "Right contains 1 more item:",
+            "{'c': 2}",
+            "Use -v to get more diff",
         ]
 
     def test_sequence_different_items(self) -> None:

